packages/@aws-cdk/cli-plugin-contract/lib/index.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * The basic contract for plug-ins to adhere to::
+ *
+ *   import { CustomCredentialProviderSource, CredentialProviderSourceRepository, Plugin } from '@aws-cdk/cli-plugin-contract';
+ *
+ *   export default class FooCDKPlugIn implements Plugin {
+ *     public readonly version = '1';
+ *
+ *     public init(host: CredentialProviderSourceRepository) {
+ *       host.registerCredentialProviderSource(new CustomCredentialProviderSource());
+ *     }
+ *   }
+ *
+ */
+export interface Plugin {
+  /**
+   * The version of the plug-in interface used by the plug-in. This will be used by
+   * the plug-in host to handle version changes.
+   */
+  version: '1';
+
+  /**
+   * When defined, this function is invoked right after the plug-in has been loaded,
+   * so that the plug-in is able to initialize itself. It may call methods of the
+   * `CredentialProviderSourceRepository` instance it receives to register new
+   * `CredentialProviderSource` instances.
+   */
+  init?: (host: CredentialProviderSourceRepository) => void;
+}
+
+export interface AwsCredentials {
+  /**
+   * AWS access key ID
+   */
+  readonly accessKeyId: string;
+
+  /**
+   * AWS secret access key
+   */
+  readonly secretAccessKey: string;
+
+  /**
+   * A security or session token to use with these credentials. Usually
+   * present for temporary credentials.
+   */
+  readonly sessionToken?: string;
+
+  /**
+   * AWS credential scope for this set of credentials.
+   */
+  readonly credentialScope?: string;
+
+  /**
+   * AWS accountId.
+   */
+  readonly accountId?: string;
+
+  /**
+   *   A Date when the identity or credential will no longer be accepted.
+   */
+  readonly expiration?:	Date;
+
+  /**
+   * Refreshes the current credentials. This function only exists for
+   * legacy reasons, to be compatible with the `AWS.Credentials` class.
+   * Plugins that use the AWS SDK v3 don't need this.
+   */
+  getPromise?: () => Promise<void>;
+}
+
+export enum Mode {
```

Comment:
I've considered making this a `const enum` but, given the [documented pitfalls], I've decided against it. This value will probably be used only once during the execution of the plugin, which itself is part of much bigger CDK command execution. So the cost of additional indirection when accessing enum values is negligible.

[documented pitfalls]: https://www.typescriptlang.org/docs/handbook/enums.html#const-enum-pitfalls

Reply:
The goal of that was more to avoid having plugins take a runtime dependency on this package. They should be able to take only a `devDependency` on this package.

What is the concrete pitfall you're concerned about?